### PR TITLE
Changing `Decoder` trait to be more composable.

### LINKED ISSUE #872: Add an `Sequence` object to the decoders
I wonder if it would be useful to have a sequence object for the decoders too.

It seems to me for example that if we build a tokenizer with a BPE model that defines a `end_of_word_suffix`, we will need to use the `BPEDecoder` decoder to replace the`end_of_word_suffix` and if we also used a ByteLevel pre-tokenization we will need the `ByteLevel` decoder to  realign the codes.

At the moment, it seems to me that we don't have a solution to choose a suitable decoder for such a tokenizer.

What do you think? :smile:

### PR DESCRIPTION
**Pull request by Narsil**
Fix #872

### CODE REVIEW

File: tokenizers/src/decoders/ctc.rs [hunk 1]
```diff
@@ -97,7 +108,13 @@ mod tests {
...
         assert_eq!(
             ctc_decoder.decode(id_to_string_result).unwrap(),
-            "HIS INSTANCT PANIC WAS FOLLOWED BY A SMALL SHARP BLOW HIGH ON HIS CHEST ".to_string()
+            vec![
+                "H", "I", "S", " ", "I", "N", "S", "T", "A", "N", "C", "T", " ", "P", "A", "N",
```
Comment (McPatate) [inline tokenizers/src/decoders/ctc.rs:112]: Instanct ?

File: tokenizers/src/decoders/ctc.rs [hunk 2]
```diff
@@ -97,7 +108,13 @@ mod tests {
...
+                "I", "C", " ", "W", "A", "S", " ", "F", "O", "L", "L", "O", "W", "E", "D", " ",
+                "B", "Y", " ", "A", " ", "S", "M", "A", "L", "L", " ", "S", "H", "A", "R", "P",
+                " ", "B", "L", "O", "W", " ", "H", "I", "G", "H", " ", "O", "N", " ", "H", "I",
+                "S", " ", "C", "H", "E", "S", "T", " "
```
Comment (McPatate) [inline tokenizers/src/decoders/ctc.rs:116]: Is the space at the end normal ?

File: tokenizers/src/decoders/wordpiece.rs [hunk 3]
```diff
@@ -44,12 +44,21 @@ pub fn cleanup(dirty_input: String) -> String {
...
+            .iter_mut()
+            .enumerate()
+            .map(|(i, token)| {
+                if token.starts_with(&self.prefix) {
+                    *token = token.replacen(&self.prefix, "", 1);
+                } else if i != 0 {
+                    *token = format!(" {}", token);
+                }
+                if self.cleanup {
+                    *token = cleanup(token.clone());
```
Comment (McPatate) [inline tokenizers/src/decoders/wordpiece.rs:58]: Isn't better API-wise to pass in a reference, `&token` and call a `.to_owned()` or `.clone()` inside the `cleanup()` method ?
Feels cleaner, the responsibility of the clone is imo `cleanup`'s.

**Review by McPatate — APPROVED** [review]
Few questions for comprehension.

Looks good overall !

File: bindings/node/lib/bindings/decoders.test.ts [hunk 4]
```diff
@@ -39,6 +39,6 @@ describe("ctcDecoder", () => {
   it("encodes correctly", () => {
     expect(
       ctcDecoder().decode(["<pad>", "h", "h", "e", "e", "l", "l", "<pad>", "l", "l", "o"])
-    ).toEqual("hello");
+    ).toEqual(["h", "e", "l", "l", "o"]);
```
[the remark this answers is not in the document]
Comment (Narsil) [reply bindings/node/lib/bindings/decoders.test.ts:42]: It dedpuplicates earlier, `<pad>` acts as a separator to  enable actually multiple letters to be outputted.

File: bindings/python/src/decoders.rs [hunk 5]
```diff
@@ -337,12 +337,12 @@ impl CustomDecoder {
...
 
 impl Decoder for CustomDecoder {
-    fn decode(&self, tokens: Vec<String>) -> tk::Result<String> {
+    fn decode(&self, tokens: Vec<String>) -> tk::Result<Vec<String>> {
         Python::with_gil(|py| {
             let decoded = self
                 .inner
                 .call_method(py, "decode", (tokens,), None)?
-                .extract::<String>(py)?;
+                .extract::<_>(py)?;
```
[the remark this answers is not in the document]
Comment (Narsil) [reply bindings/python/src/decoders.rs:345]: Well it's evidently `Vec<String>` from the return type.

I am not sure why I just didn't remove the turbofish altogether.
the other methods also just use pure `.into()` and `.extract` without turbofish.

File: tokenizers/src/decoders/ctc.rs (hunk 1 above)
Comment (Narsil) [reply tokenizers/src/decoders/ctc.rs:112]: It's the original test, I am pretty sure it just comes from the output of some wav2vec2 which actually had issues.

File: tokenizers/src/decoders/ctc.rs (hunk 2 above)
Comment (Narsil) [reply tokenizers/src/decoders/ctc.rs:116]: Yes, it was in the original test too. 

In `transformers` it's removed at the end, but the model really outputs it.

The token is actually called `word_delimiter` we just happen to use a space on space separated languages :)

File: tokenizers/src/decoders/wordpiece.rs (hunk 3 above)
Comment (Narsil) [reply tokenizers/src/decoders/wordpiece.rs:58]: I think it's almost always preferable to not clone on a user's behalf and let the user decide how best to pass owned instances (if you really require ownership).

It seems maybe this function doesn't actually require ownership, I'll change to reflect that.

File: tokenizers/src/decoders/ctc.rs (hunk 1 above)
Comment (McPatate) [reply tokenizers/src/decoders/ctc.rs:112]: Aah legacy :D

File: tokenizers/src/decoders/wordpiece.rs (hunk 3 above)
Comment (McPatate) [reply tokenizers/src/decoders/wordpiece.rs:58]: I see what you mean. W/out the clone, you will have to create a new instance of a `String`, which is somehow similar to cloning ^^

**Comment by SaulLu** [comment]
I'm sorry, I didn't have time to do a proper review. Nevertheless, I would be very interested to understand a bit better the changes that have been made :slightly_smiling_face: 

In particular, I see that the test target has changed (e.g.`.toEqual("hello");` to `.toEqual(["h", "e", "l", "l", "o"]);`): does this mean that the decode output will be different now?

**Comment by Narsil** [comment]
No not the final thing, basically the output of the tokenizer will be `"".join(decoder.decode(tokens))` if you will.

Only breaking changes are if you are using decoders directly.

**Comment by SaulLu** [comment]
Thank you very much for your answer! It's really great!

To put all the pieces together, am I right to understand that this last step is done here :smile: : https://github.com/huggingface/tokenizers/blob/2620531e9a360a4d3a010581fd28b7b1f0a5f3bb/tokenizers/src/tokenizer/mod.rs#L781

**Comment by Narsil** [comment]
Indeed !  You see Rust is not that hard ^_^

File: tokenizers/src/decoders/wordpiece.rs (hunk 3 above)
Comment (McPatate) [reply tokenizers/src/decoders/wordpiece.rs]: If that is the rust way, so be it 😄

**Pull request merged**